### DISCUSSION TITLE
Adding custom tags in PMD Sarif file from xpath rules

### DISCUSSION THREAD

**Discussion by SakshiV19** [pmd/pmd#4681] [Q&A]
Is there a way to add custom tags in PMD sarif file?

In the xpath rule, I've added a property as below:
`<property name="action" type="String" value="Document Only" description="A String Property."/>`

Once PMD runs and a sarif file is generated, is there a way to customise the 'tags' of rules in the sarif.json:
**Actual :**
`"rules": [
            {
              "id": "Debuglogs",
              "shortDescription": {
                "text": "Avoid debug statements"
              },
              "fullDescription": {},
              "helpUri": "",
              "help": {},
              "properties": {
                "ruleset": "Debug Statements Count",
                "priority": 1,
                "tags": [
                  "Debug Statements Count"
                ]
              }
            }
          ]`
          
**Expected:**
`"rules": [
            {
              "id": "Debuglogs",
              "shortDescription": {
                "text": "Avoid debug statements"
              },
              "fullDescription": {},
              "helpUri": "",
              "help": {},
              "properties": {
                "ruleset": "Debug Statements Count",
                "priority": 1,
                "tags": [
                  "Debug Statements Count",
                  "Document Only"
                ]
              }
            }
          ]`
          
If not **tags**, then any other way to customise the rule in the sarif with the custom proeprty from ruleset?

Thanks in advance.

**Comment by adangel** [comment]
The tags are currently generated from the ruleset name: https://github.com/pmd/pmd/blob/347d5992d13be82418cc1c7018bf6f7e4db7853d/pmd-core/src/main/java/net/sourceforge/pmd/renderers/internal/sarif/SarifLogBuilder.java#L185

Not sure, why we put the rulesetname as a tag, but that's how it was implemented back then.

Reading the Sarif Spec https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/sarif-v2.1.0-errata01-os-complete.html#_Toc141790700

> Tags **SHOULD NOT** be used to label a result or a rule as belonging to a category [...]. Instead, taxonomies (§3.19.3) **SHOULD** be used for this purpose.

That we put rulesetname into tags is actually wrong.
However, for me it is not clear, what tags should be used for...

Rule Properties would fit inside "3.50 reportingConfiguration object" and "3.51 configurationOverride object" as parameters - and that's probably not implemented yet on PMD side. That means, the sarif report doesn't contain the information, which property values have actually been used when PMD was executed.

But before changing anything, we should first try to understand your use case... Why do you want to add a tag? What is it used for? Where does its value come from?

Using rule properties to add tags doesn't look correct. Rule properties provide a configuration possibility for rules - tags are something else (although I actually don't know what tags are and what they potentially being used for...).

**Reply by SakshiV19** [reply]
We are exploring pumping more data in the form of tags/properties into our ruleset so we can populate more relevant info into our sarif and then categorise and process the results.

So we were looking for a way to pass a rule property/tag in sarif for easier processing rather than checking two files which would impact performance.

We are not necessarily looking at customising the **tags**, it could be anything.

https://docs.pmd-code.org/latest/pmd_userdocs_report_formats.html
This document says - _PMD holds an initial implementation version of SARIF rendering. This means SARIF allows for more complexity and the current implementation can be extended._

So we're looking for ways to extend it.
